File: settlement-service/src/test/java/github/lms/lemuel/settlement/adapter/in/web/ReconciliationControllerTest.java
package github.lms.lemuel.settlement.adapter.in.web;

import github.lms.lemuel.common.config.jwt.JwtUtil;
import github.lms.lemuel.settlement.application.port.in.ReconcileDailyTotalsUseCase;
import github.lms.lemuel.settlement.domain.ReconciliationReport;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.webmvc.test.autoconfigure.AutoConfigureMockMvc;
import org.springframework.boot.webmvc.test.autoconfigure.WebMvcTest;
import org.springframework.test.context.bean.override.mockito.MockitoBean;
import org.springframework.test.web.servlet.MockMvc;

import java.math.BigDecimal;
import java.time.LocalDate;

import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@WebMvcTest(controllers = ReconciliationController.class)
@AutoConfigureMockMvc(addFilters = false)
class ReconciliationControllerTest {

    @Autowired MockMvc mockMvc;
    @MockitoBean JwtUtil jwtUtil;
    @MockitoBean ReconcileDailyTotalsUseCase reconcileUseCase;

    @Test
    @DisplayName("GET /admin/reconciliation runs daily reconciliation")
    void run() throws Exception {
        LocalDate date = LocalDate.of(2026, 4, 1);
        when(reconcileUseCase.reconcile(date)).thenReturn(ReconciliationReport.of(
                date,
                new BigDecimal("100000"), // 캡처 gross
                new BigDecimal("10000"),  // 환불
                new BigDecimal("96500"),  // 생성된 정산 net
                new BigDecimal("3500"),   // 생성된 정산 commission (3.5%)
                new BigDecimal("10000")   // 환불 조정
        ));

        mockMvc.perform(get("/admin/reconciliation").param("date", "2026-04-01"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.targetDate").exists())
                .andExpect(jsonPath("$.discrepancy").value(0.00))
                .andExpect(jsonPath("$.matched").value(true));
    }

    @Test
    @DisplayName("GET /admin/reconciliation requires date")
    void runMissingDate() throws Exception {
        mockMvc.perform(get("/admin/reconciliation"))
                .andExpect(status().isBadRequest());
    }
}

File: reconciliation-service/src/test/kotlin/github/lms/lemuel/reconciliation/adapter/in/web/ReconciliationControllerTest.kt
package github.lms.lemuel.reconciliation.adapter.`in`.web

import github.lms.lemuel.reconciliation.domain.DiscrepancyType
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest
import org.springframework.context.annotation.Import
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.get
import org.springframework.test.web.servlet.post

import github.lms.lemuel.reconciliation.adapter.out.source.SampleActualSource
import github.lms.lemuel.reconciliation.adapter.out.source.SampleExpectedSource
import github.lms.lemuel.reconciliation.application.ReconciliationService

/**
 * Slice test for the web adapter. Imports the real service + sample sources so
 * the demo runs end-to-end through the controller.
 */
@WebMvcTest(ReconciliationController::class)
@Import(ReconciliationService::class, SampleExpectedSource::class, SampleActualSource::class)
class ReconciliationControllerTest(@Autowired val mockMvc: MockMvc) {

    @Test
    fun `demo returns a report containing every discrepancy type`() {
        mockMvc.get("/reconciliation/demo")
            .andExpect { status { isOk() } }
            .andExpect { jsonPath("$.matchedCount") { value(2) } } // pay_1001 + pay_1005
            .andExpect { jsonPath("$.byType.${DiscrepancyType.MISSING}") { value(1) } }
            .andExpect { jsonPath("$.byType.${DiscrepancyType.EXTRA}") { value(1) } }
            .andExpect { jsonPath("$.byType.${DiscrepancyType.AMOUNT_MISMATCH}") { value(1) } }
            .andExpect { jsonPath("$.byType.${DiscrepancyType.STATUS_MISMATCH}") { value(1) } }
            .andExpect { jsonPath("$.discrepancyCount") { value(4) } }
    }

    @Test
    fun `run reconciles supplied sets and honors tolerance`() {
        val body = """
            {
              "expected": [
                {"businessKey":"k1","amountKrw":1000,"status":"PAID"},
                {"businessKey":"k2","amountKrw":2000,"status":"PAID"}
              ],
              "actual": [
                {"businessKey":"k1","amountKrw":1000,"status":"PAID"},
                {"businessKey":"k2","amountKrw":2500,"status":"PAID"}
              ],
              "toleranceKrw": 1
            }
        """.trimIndent()

        mockMvc.post("/reconciliation/run") {
            contentType = MediaType.APPLICATION_JSON
            content = body
        }
            .andExpect { status { isOk() } }
            .andExpect { jsonPath("$.matchedCount") { value(1) } }
            .andExpect { jsonPath("$.discrepancyCount") { value(1) } }
            .andExpect { jsonPath("$.discrepancies[0].type") { value("AMOUNT_MISMATCH") } }
    }
}
